Relax standardness rules regarding CHECKMULTISIG

roconnor-blockstream opened the pull request:
In PR #5247, the STRICTENC standardness rules were tightened with regards to CHECKMULTISIG so that unparsable public keys fail the script when they are encountered. The overall purpose here was to disallow the use of confusing hybrid public keys by policy while keeping policy compatible (i.e., strictly stronger) with consensus rules.

Comments in PR #5247 note that "I don't believe it should affect any system in production"; however, this belief is/was false. Counterparty was stuffing data blobs into multisig pubkey lists, but these UTXOs were meant to be spendable because, although some pubkeys were unparsable, other keys were parsable, and the UTXOs were meant to be spent by those valid keys.

But in tackling the hybrid key issue, PR #5247 disallowed any unparsable keys in multisigs, whether or not they were hybrid, and whether or not the signature was meant to satisfy a hybrid key.

In production, Counterparty UTXOs were inadvertently caught up in this standardness rule change and became "soft confiscated."  That is, they were no longer spendable by policy but still recoverable if users are able to somehow bypass standardness by mining their transactions themselves, or getting out-of-band assistance from some other miner.

I understand that Bitcoin Core never intended to "soft confiscate" any UTXOs by policy changes.  This change addresses the problem.

With this change, only hybrid formatted keys and empty keys will cause early abort in CHECKMULTISIG operations. Otherwise, failing signature checks revert to their consensus behaviour of testing subsequent public keys.

Counterparty UTXOs were never intended to make use of hybrid keys and thus shouldn't have any passing signatures using hybrid keys.

instagibbs commented:
If anyone has documentation showing the format (pre-OP_RETURN) this would be useful.

ajtowns commented:
This change affects enforcement of `SCRIPT_VERIFY_WITNESS_PUBKEYTYPE` which afaics should not be weakened [bip143](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki#restrictions-on-public-key-type)?

ajtowns commented:
If you're deliberately changing the standardness rules specified in BIP143, that should be in the PR description, and possibly proposed as an update to the BIP. If you're not deliberately changing them, they shouldn't be changed by accident. Personally, I don't see the justification for changing them here, though I don't really understand why it was specced as a policy rule rather than a consensus one at the time, either.

Personally, I don't really understand the motivation for reconsidering a policy rule that's been in place for 11 years, either, and for which new adoption would only be less fee efficient than what's already allowed. If there were real coins locked due to this policy, it seems like they should have been already recovered via getting a non-standard spend mined directly.

ajtowns commented:
> > If there were real coins locked due to this policy, it seems like they should have been already recovered via getting a non-standard spend mined directly.
> 
> I don't know why you would think that they should have already been recovered. You think anyone can just call up a mining pool and get their non-standard transactions mined?

That's what things like slipstream allow today, so, yes? Template building by hashers also make that more accessible today than previously. In any event, if it's not worth calling up a single miner and getting it mined, it also doesn't seem worth changing the default policy for every miner that follows our defaults.

> How much money needs to be at stake before you can even get a pool's attention? Or how small in value do UTXOs have to be so that it is okay to soft-confiscate it since it isn't worth fixing for such a little amount?

I don't think chastising the reader about confiscating coins is constructive. If you want to make those coins easily accessible now, that's a fine thing to argue for, but it should be done on its merits, not by attempting to guilt trip people. This PR isn't passing CI, for instance.

> Anyhow, AFAIU Counterparty was using bare multisigs, so in theory we don't have to speculate and we can just look for such UTXOs that are waiting to be freed.

There seem to be six utxos from prior to block 400,000 that currently have a value of more than $100 USD that fit the description here:

 * [176be48b7e1091915b48c12c01cfeb8e301dba09988c49314e696906ca383cf2:0](https://mempool.space/tx/176be48b7e1091915b48c12c01cfeb8e301dba09988c49314e696906ca383cf2) - 2014-03-06 960k sat / $1050
 * [3482ea1d219120b13f3be49f63fd7079136785dff2bf7487726b640da5876e33:0](https://mempool.space/tx/3482ea1d219120b13f3be49f63fd7079136785dff2bf7487726b640da5876e33) - 2014-09-11 2.0M sat / $2190
 * [d6d89928cf8cf92ae3db54898329174c9cde766f63a8aa0cca6da6d17bcaf17d:0](https://mempool.space/tx/d6d89928cf8cf92ae3db54898329174c9cde766f63a8aa0cca6da6d17bcaf17d) - 2015-02-13 3.8M sat / $4160
 * [2cc8dcd16da3912e54429946eecc5f7f1755ea3bbbd0ea60e5906209c73d6b14:0](https://mempool.space/tx/2cc8dcd16da3912e54429946eecc5f7f1755ea3bbbd0ea60e5906209c73d6b14) - 2015-02-20 500k sat / $547
 * [28a21c72d59649744a8c93683eb5d28a14e5ebcfa1252d63553a81dcabdda140:0](https://mempool.space/tx/28a21c72d59649744a8c93683eb5d28a14e5ebcfa1252d63553a81dcabdda140) - 2015-03-28 100k sat / $109
 * [97ae8b20c783ed6deecc2a4578a0c36861296ff40146fb3cc9b10f0944dca960:0](https://mempool.space/tx/97ae8b20c783ed6deecc2a4578a0c36861296ff40146fb3cc9b10f0944dca960) - 2015-10-02 200k sat / $219
 
The 3.8M sat output was worth about $11 at the time, so I could quite easily see it having been written off as a loss and access to the corresponding private keys not having been maintained in the meantime.

There are an additional 15,133 outputs from prior to block 400k that look like they might fit this case, all with between 780 and 30,919 sats, totalling 1.0148,8939 BTC, which had a total value of about $472 if I pick the highest BTC price in the above timeframe. (Or at least there were as at block 897666 which is the utxo data I have handy for analysing. This likely includes some false positives as well, though I think I eliminated most things)

For comparison, it looks to me like there was 0.9068,0410 BTC (so $422 at the time by the same logic) across perhaps 11k bare multisig addresses that can only be spent via uncompressed pubkeys (eg [this](https://mempool.space/tx/7af3670ba59cd6e11e0a36149dcc2042e16e5d86d155ca8bbdba1beb59625feb) looks like ~10M sats worth ~$10k USD today, sent to a 2/3 uncompressed bare multisig, from the same multisig, so at one point certainly spendable).

> Soft-confiscating these transactions was wrong, and we should fix mistakes if we can when we realize we have made one, [...]

Soft-confiscating is always a risk with policy restrictions, but there is also already a straight-forward solution to all problems in that class: talk to a miner directly.

roconnor-blockstream commented:
Those outputs look like multisigs that are composed out of valid compressed or uncompressed pubkeys, and I believe spending them would pass standardness rules (except for 3482ea1d219120b13f3be49f63fd7079136785dff2bf7487726b640da5876e33:0, which appears to simply be a broken script).

The UTXO's I'm looking for would have pubkeys with an invalid prefix, or an invalid length for their prefix.

Based on the code prior to https://github.com/CounterpartyXCP/counterparty-core/pull/443/commits/f1a8c82914b2f913560496e3d1becc1b63f88118 I would specifically expect a 1-of-3 multisig where at least 1 of the keys is 33 bytes but does not begin with 0x02 or 0x03.

roconnor-blockstream committed:
Relax standardness rules regarding CHECKMULTISIG

In PR #5247, the STRICTENC standardness rules were tightened with regards to
CHECKMULTISIG so that unparsable public keys fail the script when they are
encountered. The overall purpose here was to disallow the use of confusing
hybrid public keys by policy while keeping policy compatible (i.e., strictly
stronger) with consensus rules.

Comments in PR #5247 note that "I don't believe it should affect any system in
production"; however, this belief is/was false. Counterparty was stuffing data
blobs into multisig pubkey lists, but these UTXOs were meant to be spendable
because, although some pubkeys were unparsable, other keys were parsable, and
the UTXOs were meant to be spent by those valid keys.

But in tackling the hybrid key issue, PR #5247 disallowed any unparsable keys in
multisigs, whether or not they were hybrid, and whether or not the signature was
meant to satisfy a hybrid key.

In production, Counterparty UTXOs were inadvertently caught up in this
standardness rule change and became "soft confiscated." That is, they were no
longer spendable by policy but still recoverable if users are able to somehow
bypass standardness by mining their transactions themselves, or getting
out-of-band assistance from some other miner.

I understand that Bitcoin Core never intended to "soft confiscate" any UTXOs by
policy changes. This change addresses the problem.

With this change, only hybrid formatted keys and empty keys will cause early
abort in CHECKMULTISIG operations. Otherwise, failing signature checks revert to
their consensus behaviour of testing subsequent public keys.

Counterparty UTXOs were never intended to make use of hybrid keys and thus
shouldn't have any passing signatures using hybrid keys.

roconnor-blockstream commented:
At the risk of opening the bikeshed, another reasonable option I was thinking of today is allowing arbitrary data up to 33 bytes, or an uncompressed pubkey:

    bool static IsDataOrUncompressedPubKey(const valtype &vchPubKey) {
        if (!vchPubKey.empty() && vchPubKey.size() <= CPubKey::COMPRESSED_SIZE) {
            // A "data" push or a compressed pubkey.
            return true;
        }
        if (vchPubKey[0] == 0x04 && vchPubKey.size() == CPubKey::SIZE) {
            // An uncompressed pubkey.
            return true;
        }
        return false;
    }

I'm not sure if any of the 147,470 UTXO would pass one condition but not the other.

For the moment I've pushed @ajtowns's suggested policy.  I'm fine with that one or the one above or anything in between.

luke-jr commented:
Concept NACK, they "soft confiscated" their own coins. This isn't a good excuse to enable spam.

ajtowns commented:
> At the risk of opening the bikeshed, another reasonable option I was thinking of today is allowing arbitrary data up to 33 bytes, or an uncompressed pubkey:

I see two misses for this approach (though I'm still not looking at any utxos after block 400k):

 * [7be22151bc96041932830d79c2af4472e40e9dedaa29018fff6757501e7855c6:0](https://mempool.space/tx/7be22151bc96041932830d79c2af4472e40e9dedaa29018fff6757501e7855c6) 20000 sats, block 344398 (two 65 byte data pushes on top of a compressed key)
 * [15595590636afdf807462551fb58655a18419cfdb9a0ad9c866cf00f74108bc5:1](https://mempool.space/tx/15595590636afdf807462551fb58655a18419cfdb9a0ad9c866cf00f74108bc5) 10860 sats, block 280515 (probably a compressed key mistakenly pushed with an extra 00 byte on the end, making for an invalid pubkey, on top of a compressed and an uncompressed key)

(EDIT: also, keep `vchPubKey.empty()` as a separate test, both for readability (I missed it!) and to that `vchPubKey[0] == 0x04` doesn't fault...)

ajtowns commented:
It would be good to add a test case that demonstrates scriptPubKeys of this form (baremultisig and possible multisig in p2sh, with valid key push prior to invalid key) are spendable.

I think creating these outputs became non-standard with #12460 (2018-04-05 merge date), which changed the checks for pubkeys in a bare multisig from a size between 33 and 65 bytes (inclusive) to an explicit check of 02/03-and-33-bytes or 04/06/07-and-65-bytes.

In trying to create a functional test to validate these become spendable, I'm getting `bad-txns-nonstandard-inputs` which indicates they're still unspendable, because `Solver` doesn't consider them to be a sane multisig scriptPubKey.

Fix and test case at https://github.com/ajtowns/bitcoin/commits/202511-cparty-spends/ Not really convinced adding params to Solver is sensible, but also pulling the pubkey size checks out also seems silly.

ajtowns commented:
> Why doesn't `tx.get_vsize()` return the correct value in the code above. Did I make a stupid mistake somewhere?

The vsize that `check_mempool_result()` (aka `testmempoolaccept`) returns is adjusted for the sigop count (which is treated as 20 due to `fAccurate=false` via `GetTransactionSigOpCost()`/`GetLegacySigOpCount()`), while `get_vsize` is not. See also #32800.

Christewart commented:
> It would be good to add a test case that demonstrates scriptPubKeys of this form (baremultisig and possible multisig in p2sh, with valid key push prior to invalid key) are spendable.

Hi AJ, are these the test vectors you had in mind? 

```
[
    "0 0x47 0x3044022044dc17b0887c161bb67ba9635bf758735bdde503e4b0a0987f587f14a4e1143d022009a215772d49a85dae40d8ca03955af26ad3978a0ff965faa12915e9586249a501",
    "1 0 0x21 0x02865c40293a680cb9c020e7b1e106d8c1916d3cef99aa431a56d253e69256dac0 2 CHECKMULTISIG NOT",
    "P2SH,STRICTENC",
    "PUBKEYTYPE",
    "1-of-2 CHECKMULTISIG NOT with the first pubkey invalid, with a 2nd valid signature/key combo."
],
[
    "0 0x47 0x3044022044dc17b0887c161bb67ba9635bf758735bdde503e4b0a0987f587f14a4e1143d022009a215772d49a85dae40d8ca03955af26ad3978a0ff965faa12915e9586249a501",
    "1 0 0x21 0x02865c40293a680cb9c020e7b1e106d8c1916d3cef99aa431a56d253e69256dac0 2 CHECKMULTISIG NOT",
    "P2SH",
    "OK",
    "1-of-2 CHECKMULTISIG NOT with the first pubkey invalid, with a 2nd valid signature/key combo."
]
```

You can run them here with `script_tests.json` minimized to relevant test cases (I believe): https://github.com/bitcoin/bitcoin/commit/c41a23dcf224c10cdbcb367c0e5495df34538bb4 . I can PR if I'm on the right path

sedited commented:
@pinheadmz do you want to take a look here?

pinheadmz commented:
I spent a few days reviewing all the commit history and discussion here and in https://github.com/bitcoin/bitcoin/issues/33882. I've been able to contextualize just about all of it, but I need to revisit the effect it has on bip143.

Anyway, I took a utxo snapshot and analyzed it with claude to determine how much utxo could be rescued by the policy change: https://gist.github.com/pinheadmz/e599afb5f926131ae5d9c67ef68ce09b

Seems like the bulk is around 80,000 UTXOs worth $1-10 each and consuming maybe around 10 MB of utxo set space total. So it's certainly not an urgent issue. I even have a 2015-era counterparty wallet around here somewhere... but even if I did dig it up I don't think I'd personally bother scraping the loose change.

I generally agree with @roconnor-blockstream that #5247 was overkill and its something we should fix. Probably the most important motivation is in #33882  just clarifying and sane-ifying the multisig policy (making spend match create, etc).

As a PR, this could use tests before and after the change which I think @151henry151 might have already written, but it would definitely help understand all the implications.

151henry151 commented:
@pinheadmz The tests I wrote are in #33926 (closed) — `AreInputsStandard` on non-solvable legacy scripts. They don’t exercise `STRICTENC` or spending the counterparty-style outputs this PR is about. `WITNESS_PUBKEYTYPE` isn’t touched, and anything that would newly get past `STRICTENC` still fails `IsCompressedPubKey`, so I wouldn’t expect v0 witness acceptance to change. The UTXOs the gist counts as rescued would still hit `bad-txns-nonstandard-inputs` because `Solver` wants `ValidSize` keys; this only stops the `STRICTENC` abort, and task 2 in #33882 is the other piece. AJ has a counterparty-style spend test on ajtowns/bitcoin `202511-cparty-spends` — I can adapt that into before/after coverage here.
